### PR TITLE
Feature/bitbucket filter

### PR DESCRIPTION
**Pull request by adelkahomolova**
Add possibility to filter pullrequests.

### CODE REVIEW

File: src/inspectors/CollaborationInspector.ts [hunk 1]
```diff
@@ -11,8 +14,12 @@ export class CollaborationInspector implements ICollaborationInspector {
     this.service = service;
   }
 
-  async getPullRequests(owner: string, repo: string) {
-    return this.service.getPullRequests(owner, repo);
+  async getPullRequests(
+    owner: string,
+    repo: string,
+    options?: ListGetterOptions<{ state?: GitHubPullRequestState | BitbucketPullRequestState }>,
```
Comment (prokopsimek) [inline src/inspectors/CollaborationInspector.ts]: use our own `PullRequestState`

File: src/services/bitbucket/BitbucketService.ts [hunk 2]
```diff
@@ -73,46 +75,59 @@ export class BitbucketService implements ICVSService {
...
-          id: val.destination.repository.uuid,
-          owner: {
-            login: <string>val.destination.repository.full_name.split('/').shift(),
-            id: <string>(await this.client.users.get({ username: `${val.destination.repository.full_name.split('/').shift()}` })).data.uuid
-              ? <string>(await this.client.users.get({ username: `${val.destination.repository.full_name.split('/').shift()}` })).data.uuid
-              : 'undefined',
-            url: url.concat(`/${val.destination.repository.full_name.split('/').shift()}`),
+    const values = response.data.values.map(async (val) => {
+      const ownerId = <string>(
+        (await this.client.users.get({ username: `${val.destination.repository.full_name.split('/').shift()}` })).data.uuid
```
Comment (prokopsimek) [inline src/services/bitbucket/BitbucketService.ts]: don't call it in a map

File: src/services/git/GitHubService.ts [hunk 3]
```diff
@@ -71,7 +70,7 @@ export class GitHubService implements ICVSService {
   async getPullRequests(
     owner: string,
     repo: string,
-    options?: ListGetterOptions<{ state?: GitHubPullRequestState }>,
+    options?: ListGetterOptions<{ state?: GitHubPullRequestState | BitbucketPullRequestState }>,
```
Comment (prokopsimek) [inline src/services/git/GitHubService.ts]: - `BitbucketPullRequestState `

File: src/services/git/ICVSService.ts [hunk 4]
```diff
@@ -1,13 +1,13 @@
...
 import { Paginated } from '../../inspectors/common/Paginated';
 import { GitHubPullRequestState } from './IGitHubService';
+import { Commit, Contributor, ContributorStats, Directory, File, Issue, PullFiles, PullRequest, PullRequestReview, Symlink } from './model';
 
 export interface ICVSService {
   getPullRequests(
     owner: string,
     repo: string,
-    options?: ListGetterOptions<{ state?: GitHubPullRequestState }>,
+    options?: ListGetterOptions<{ state?: GitHubPullRequestState | BitbucketPullRequestState }>,
```
Comment (prokopsimek) [inline src/services/git/ICVSService.ts]: `PullRequestState`

File: src/services/git/GitHubService.ts [hunk 5]
```diff
@@ -71,11 +71,12 @@ export class GitHubService implements ICVSService {
...
     owner: string,
     repo: string,
-    options?: ListGetterOptions<{ state?: GitHubPullRequestState }>,
+    options?: ListGetterOptions<{ state?: PullRequestState }>,
   ): Promise<Paginated<PullRequest>> {
     let url = 'GET /repos/:owner/:repo/pulls';
     if (options !== undefined && options.filter !== undefined && options.filter.state !== undefined) {
-      url = `${url}?state=${options.filter.state}`;
+      const state = CSVServicesUtils.getPRState(options.filter.state, CSVService.github);
+      url = `${url}?state=${state}`;
```
Comment (prokopsimek) [inline src/services/git/GitHubService.ts]: Do not build URL as a string... It'll be a problem when we have 2 or more query parameters. Use `qs` library - https://www.npmjs.com/package/qs#stringifying

File: src/services/git/CSVServicesUtils.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+  ): GitHubPullRequestState | BitbucketPullRequestState | undefined => {
+    if (service === CSVService.github) {
+      switch (state) {
+        case PullRequestState.open:
+          return GitHubPullRequestState.open;
+        case PullRequestState.closed:
+          return GitHubPullRequestState.closed;
+        case PullRequestState.all:
+          return GitHubPullRequestState.all;
+        case undefined:
```
Comment (prokopsimek) [inline src/services/git/CSVServicesUtils.ts]:
```suggestion
        default:
```

File: src/services/git/CSVServicesUtils.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+    if (service === CSVService.bitbucket) {
+      switch (state) {
+        case PullRequestState.open:
+          return BitbucketPullRequestState.open;
+        case PullRequestState.closed:
+          return BitbucketPullRequestState.closed;
+        case PullRequestState.all:
+          return BitbucketPullRequestState.open && BitbucketPullRequestState.closed && BitbucketPullRequestState.declined;
+        case undefined:
```
Comment (prokopsimek) [inline src/services/git/CSVServicesUtils.ts]:
```suggestion
        default:
```

File: src/services/git/CSVServicesUtils.ts [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    if (service === CSVService.bitbucket) {
+      switch (state) {
+        case PullRequestState.open:
+          return BitbucketPullRequestState.open;
+        case PullRequestState.closed:
+          return BitbucketPullRequestState.closed;
+        case PullRequestState.all:
+          return BitbucketPullRequestState.open && BitbucketPullRequestState.closed && BitbucketPullRequestState.declined;
```
Comment (prokopsimek) [inline src/services/git/CSVServicesUtils.ts]: WTF? 🤣

Comment (prokopsimek) [reply src/services/git/CSVServicesUtils.ts]: This won't work. Write tests for this utils.

File: test/helpers/bitbucketNock.ts [hunk 9]
```diff
@@ -11,10 +11,13 @@ export class BitbucketNock {
     this.url = 'https://api.bitbucket.org/2.0';
   }
 
-  getApiResponse(resource: string, id?: number, value?: string): nock.Scope {
+  getApiResponse(resource: string, id?: number, value?: string, state?: string): nock.Scope {
     let url = `${this.url}/repositories/${this.user}/${this.repoName}/${resource}`;
     let response;
 
+    let params = {};
+    const persist = true;
```
[the remark this answers is not in the document]
Comment (adelkahomolova) [reply test/helpers/bitbucketNock.ts]: See line 41 for the params and 59 for the persist. Actually I don' change the persist but when I will, I can do it easily.

File: src/services/bitbucket/BitbucketService.ts [hunk 10]
```diff
@@ -73,46 +78,57 @@ export class BitbucketService implements ICVSService {
...
+    const stateForUri = qs.stringify({ state: state }, { addQueryPrefix: true, indices: false, arrayFormat: 'repeat' });
+    let apiUrl = `https://api.bitbucket.org/2.0/repositories/${owner}/${repo}/pullrequests`;
+    if (stateForUri) {
+      apiUrl = apiUrl.concat(`${stateForUri}`);
+    }
 
-    const response = <DeepRequired<Bitbucket.Response<Bitbucket.Schema.PaginatedPullrequests>>>await this.client.pullrequests.list(params);
-    const url = 'www.bitbucket.org';
+    const ownerUrl = `www.bitbucket.org/${owner}`;
+    const ownerId = String((await this.client.users.get({ username: owner })).data.uuid);
```
Comment (prokopsimek) [inline src/services/bitbucket/BitbucketService.ts]:
```suggestion
    const ownerId =`${await this.client.users.get({ username: owner })).data.uuid}`;
```

File: src/services/git/__MOCKS__/bitbucketServiceMockFolder/getPullRequestsResponse.ts [hunk 11]
```diff
@@ -33,3 +66,94 @@ export const getPullRequestsResponse: Paginated<PullRequest> = {
   page: 1,
   perPage: 1,
 };
```
Comment (prokopsimek) [inline src/services/git/__MOCKS__/bitbucketServiceMockFolder/getPullRequestsResponse.ts]: DRY!

File: test/helpers/bitbucketNock.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (prokopsimek) [reply test/helpers/bitbucketNock.ts]: Ok, thanks :)

File: test/helpers/bitbucketNock.ts [hunk 12]
```diff
@@ -77,22 +111,48 @@ export class BitbucketNock {
     }
     return interceptor;
   }
+
+  mockBitbucketPullRequestResponse(states: BitbucketPullRequestState | BitbucketPullRequestState[]): Paginated<PullRequest> {
```
Comment (prokopsimek) [inline test/helpers/bitbucketNock.ts]:
```suggestion
  mockBitbucketPullRequestsResponse(states: BitbucketPullRequestState | BitbucketPullRequestState[]): Paginated<PullRequest> {
```

**Pull request merged**